packages/foundation/src/slots.test.tsx

```diff
@@ -2,7 +2,115 @@ import * as React from 'react';
 import * as renderer from 'react-test-renderer';
 import { mount } from 'enzyme';
 import { withSlots, createFactory, getSlots } from './slots';
-import { IProcessedSlotProps, ISlot, ISlotProp, ISlotPropRenderFunction, ISlotDefinition } from './ISlots';
+import { IHTMLElementSlot, IHTMLSlot } from './IHTMLSlots';
+import { ExtractProps, ExtractShorthand, IProcessedSlotProps, ISlot, ISlotDefinition, ISlotProp } from './ISlots';
+
+describe('typings', () => {
+  type ITestProps = { testProp?: number };
+  const TestProps = { testProp: 42 };
+  const TestOptions = { props: TestProps };
+  // TODO: this should be able to support returning nulls
+  // const TestRender =
+  //   (props: ITestProps, defaultComponent: React.ComponentType<ITestProps>): ReturnType<React.StatelessComponent> => null;
+  const TestRender = (props: ITestProps, defaultComponent: React.ComponentType<ITestProps>) => <div />;
+  const TestComponent: React.FunctionComponent<ITestProps> = () => null;
+
+  it('do not generate TS compile errors on getSlots usage', () => {
+    interface IComponentSlots {
+      testHtmlSlot?: IHTMLSlot;
+      testElementSlot?: IHTMLElementSlot<'button'>;
+      testSlot?: ISlotProp<ITestProps, string>;
+      testSlotNoShorthand?: ISlotProp<ITestProps>;
+    }
+
+    interface IComponentProps extends IComponentSlots {
+      someProp?: number;
+    }
+
+    getSlots<IComponentProps, IComponentSlots>(
+      {},
+      {
+        testHtmlSlot: 'div',
+        testElementSlot: 'button',
+        testSlot: TestComponent,
+        testSlotNoShorthand: TestComponent
+      }
+    );
+  });
+
+  it('do not generate TS compile error on valid ISlotProp assignments', () => {
+    const p00: ISlotProp<ITestProps> = TestOptions;
```

Comment:
nit: You start this count with `p00` whereas you start the tests below with `p10`, can we change this to have a consistent approach?